bootstore/src/coordinator.rs

```diff
@@ -0,0 +1,427 @@
+/// The coordinatore is responsible for creating the [`trust_quorum::RackSecret`]
+/// splitting it into share sand distributing those shares along with the
+/// relevant information to the participant nodes.
+pub struct Coordinator {
+    state: CoordinatorState,
+    op: Box<dyn CoordinatorOperation>,
+}
+
+impl Coordinator {
+    /// Create a coordinator used to initialize a rack
+    pub fn new_initialize(
+        log: Logger,
+        rack_uuid: Uuid,
+        members: BTreeSet<Ed25519Certificate>,
+    ) -> Result<Coordinator, Error> {
+        let log = log.new(o!(
+            "component" => "BootstoreCoordinator"
+        ));
+        let total_nodes = members.len();
+        let threshold = Self::threshold(total_nodes);
+        let secret = RackSecret::new();
+        let (shares, verifier) = secret
+            .split(threshold, total_nodes)
+            .map_err(Error::FailedToSplitSecret)?;
+        let share_distributions = members
+            .iter()
+            .cloned()
+            .zip(shares.into_iter().map(|share| ShareDistribution {
+                threshold,
+                verifier: verifier.clone(),
+                share,
+                member_device_id_certs: members.clone(),
+            }))
+            .collect();
+        let state = CoordinatorState {
+            id: 0,
+            log,
+            rack_uuid,
+            total_nodes,
+            ackd_prepares: BTreeSet::new(),
+            ackd_commits: BTreeSet::new(),
+        };
+        Ok(Coordinator {
+            state,
+            op: Box::new(InitializeOperation { members, share_distributions }),
+        })
+    }
+
+    /// Handle responses from nodes.
+    ///
+    /// Return `Ok(true)` if the transaction is complete, `Ok(false)` if
+    /// the response was handled fine, but the transaction is not complete.
+    pub fn handle(
+        &mut self,
+        from: Ed25519Certificate,
+        rsp: NodeResponse,
+    ) -> Result<bool, Error> {
+        // TODO: Should  these conditions be passed up as errors?
+        // If we pass them up, should we also log them?
+        if rsp.version != 1 {
+            error!(
+                self.state.log,
+                "Invalid version for response: {}", rsp.version
+            );
+        }
+
+        if rsp.coordinator_id != self.state.id {
+            error!(
+                self.state.log,
+                "Invalid Coordinator ID. Expected: {}, Actual: {}",
+                self.state.id,
+                rsp.coordinator_id
+            );
+        }
+
+        match rsp.result {
+            Ok(result) => self.op.handle(&mut self.state, from, result),
+            Err(err) => Err(Error::NodeResponseError(err)),
+        }
+    }
+
+    /// Return a set of [`NodeRequests`] to send to all members that haven't acked yet.
+    /// A higher "networking" layer is responsible for sending these messages over
+    /// sprockets channels and receiving responses to be handled by the `Coordinator`.
```

Comment:
This is pretty much spot on except that the higher layer actually creates either an `initializing` or `reconfiguring` coordinator through a `new_XXX` method. The upper layer passes in the membership not the total number of nodes. The coordinator itself knows the protocol and what messages need sending to each Node based on what responses it has received so far in its `handle` method.  Based on this, yes, the coordinator will always return up to "total nodes" entries, and the first call is guaranteed to return exactly "total nodes" entries. A map is returned so that the upper layer can indeed map the certs to specific hosts/sprockets channels and send messages to the right destinations.

Reply:
The whole point of this split is to allow advanced property based testing and protocol implementation without worrying about async code or transport at all.